### PR TITLE
osd/OSDMap: make pg_temp more efficient

### PR DESCRIPTION
**Pull request by liewegas**
On an OSDMap with 117012 remapped pgs, we go from 17.2MB -> 6.0MB RAM, and we are 3.5x faster at decoding.

We are also 3.7x faster at mapping all PGs in the osdmap (~50% of which are remapped).

### CODE REVIEW

File: src/osd/OSDMap.h [hunk 1]
```diff
@@ -115,6 +117,154 @@ ostream& operator<<(ostream& out, const osd_xinfo_t& xi);
...
+    ::encode(n, bl);
+    for (auto &p : map) {
+      ::encode(p.first, bl);
+      bl.append((char*)p.second, (*p.second + 1) * sizeof(int32_t));
+    }
+  }
+  void decode(bufferlist::iterator& p) {
+    data.clear();
+    map.clear();
+    int32_t n;
```
Comment (tchaikov) [inline src/osd/OSDMap.h]: will continue the review later on.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-26_06:56:21-rados-wip-sage-testing---basic-smithi/
tests out ok, although i'm seeing a (possibly related) crash from mon on bigbang.

File: src/osd/OSDMap.h [hunk 2]
```diff
@@ -114,6 +116,240 @@ WRITE_CLASS_ENCODER(osd_xinfo_t)
...
+  void rebuild() {
+    bufferlist bl;
+    encode(bl);
+    auto p = bl.begin();
+    decode(p);
+  }
+  friend bool operator==(const PGTempMap& l, const PGTempMap& r) {
+    return
+      l.map.size() == r.map.size() &&
+      l.data.contents_equal(r.data);
```
Comment (tchaikov) [inline src/osd/OSDMap.h]: no need to check `l.map.size() == r.map.size()`, i think. `l.data.contents_equal(r.data)` implies `l.map.size() == r.map.size()`.

File: src/osd/OSDMap.h [hunk 3]
```diff
@@ -114,6 +116,240 @@ WRITE_CLASS_ENCODER(osd_xinfo_t)
...
+    size_t need = sizeof(int32_t) * (1 + v.size());
+    if (need < data.get_append_buffer_unused_tail_length()) {
+      bufferptr z(data.get_append_buffer_unused_tail_length());
+      z.zero();
+      data.append(z.c_str(), z.length());
+    }
+    ::encode(v, data);
+    map[pgid] = (int32_t*)(data.back().end_c_str()) - (1 + v.size());
+  }
+  mempool::osdmap::vector<int32_t> get(pg_t pgid) {
```
Comment (tchaikov) [inline src/osd/OSDMap.h]: return const ref

File: src/osd/OSDMap.h [hunk 4]
```diff
@@ -114,6 +116,240 @@ WRITE_CLASS_ENCODER(osd_xinfo_t)
...
+    return map.count(pgid);
+  }
+  void erase(pg_t pgid) {
+    map.erase(pgid);
+  }
+  void clear() {
+    map.clear();
+    data.clear();
+  }
+  void set(pg_t pgid, const mempool::osdmap::vector<int32_t>& v) {
```
Comment (tchaikov) [inline src/osd/OSDMap.h]: could have a `void set(pg_t pgid, const mempool::osdmap::vector<int32_t>&& v)` if possible.

**Comment by liewegas** [comment]
passed a qa run, and looks ok on bigbang.

**Comment by tchaikov** [comment]
@liewegas i think i've finished the review.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-29_22:34:44-rados-wip-sage-testing---basic-smithi/

**Comment by liewegas** [comment]
oops i forget to push my update

File: src/osd/OSDMap.h (hunk 3 above)
Comment (tchaikov) [inline src/osd/OSDMap.h]: i was wrong. it's a local/temp var.

**Pull request merged**